test/utils.py

```diff
@@ -0,0 +1,373 @@
+    """fn must be pickleable.
+
+    debug:
+      - 0: normal
+      - 1: print pid on start
+      - 2: also wait 15s in process to allow time to 'Attach to Process' in Pycharm """
+    dest = mp.Queue()
+    # The daemon's main purpose is to stop ourselves accidentally creating a fork-bomb
+    #  by running run_with_timeout in the worker recursively
+    #  (we do weird stuff to get pickle to work which might break?).
+    # This is because daemonic processes cannot spawn more processes (even daemonic ones).
+    # The worker exiting when we get a Ctrl+C is just an added (but necessary) bonus.
+    p = mp.Process(target=_worker_process, args=(fn, dest, args, kwargs, debug), daemon=True)
+    p.start()
+    p.join(timeout_sec)
+    if p.is_alive():
+        p.kill()  # Die!
+        raise TestTimeout(f"Function took too long (exceeded timeout of {timeout_sec}s)")
+    if p.exitcode != 0:
+        raise mp.ProcessError("Error in process") from dest.get()
+    return dest.get()
+
+
+async def join_async(p, timeout: float, interval: float = 0.005, join_timeout: float = 0.0):
+    start = time.perf_counter()
+    while p.is_alive() and time.perf_counter() < start + timeout:
+        await asyncio.sleep(interval)
+        p.join(join_timeout)
+
+
+def _pool_worker(send_my_pid: mp.managers.ValueProxy[int], fn, args, kwargs, debug):
+    send_my_pid.set(os.getpid())  # Send it so main process knows what to kill
+    return fn(*args, **kwargs)  # No exception handling - pool does it for us
+
+
+def _pool_worker_2(fn, args, kwargs, debug=0):
+    if debug >= 1:
+        print(f"Worker process running on pid={os.getpid()}")
+    if debug >= 2:
+        time.sleep(15)
+        print('After sleep')
+    return fn(*args, **kwargs)  # No exception handling - pool does it for us
+
+
+async def _run_with_timeout_async_pool(pool: SimpleProcessPool, timeout: float, target,
+                                       args=(), kwargs=None, debug=0):
+    try:
+        return await pool.apply(_pool_worker_2, (target, args, kwargs, debug),
+                                {}, timeout)
+    except TimeoutError as e:
+        raise TimeoutError(f"Function took too long "
+                           f"(exceeded timeout of {timeout}s)") from e
+    except Exception as e:
+        raise mp.ProcessError("Error in process") from e
+
+
+@dataclass
+class _Task:
+    inputs: tuple[int, Callable, tuple, dict[str, ...]]
+    output: tuple[int, bool, object | Exception] = None
+
+
+class SimpleProcessPool:
+    def __init__(self, processes: int = None):
+        self.n_processes = processes or os.cpu_count() or 4
+        self.processes: list[_ProcessWrapper] = self._create_processes()
+        self.key = 0
+        self.tasks: dict[int, _Task] = {}
+
+    def _create_processes(self):
+        return [_ProcessWrapper(self, i)
+                for i in range(self.n_processes)]
+
+    def _kill_and_restart(self, i: int):
+        self.processes[i].p.kill()
```

Comment:
`_ProcessWrapper` should probably have a `.kill()` so we don't have to touch its attributes